Make elements clickable in the images of the HTML-report

aschlechter opened the pull request:
Based on map-areas generated in the html:
-> map-areas need to be adapted when image is resized -> some javascipt has been added to the template account for this
-> there is some fiddling to get the right offsets for the elements in the image. The implementation is minimally invasive at least from the code modifications point of view. It might be possible to improve this (remove the "doubled" calls to first create the images, then to find the offsets in the images), but code changes would spread further.

Phillipus commented:
Hi, thanks for this. This is certainly worth incorporating. A couple of things:

- I'm in Paris for presentations for most of next week so can't merge this until after that 
- I'd like to see if the create image and get offset routines could be combined into one routine. As you say, it's doubled up at the moment, and these methods can be expensive

Phil

aschlechter commented:
Hi, 
I added a wrapper object to return both the image and the bounds-rectangle. I named it "GeoReferencedImage" as an analogy to georeferenced images from the gps-world (an image with the necessary information to map it to a coordinate system). Maybe there's a better name.
Changed the "working" methods in DiagramUtil to private, renamed them and have them return the wrapper object.
Added new public methods with the orginal names to redirect to the private methods while only returning the bare image.
Changed the one "private" method used in Report plugin to public.
Resulting code modifications are still quite overseeable.

Have a nice stay in Paris. 

Tun

Phillipus commented:
Thanks. Bear with me for a few days until I can spend some time reviewing it...

Phil

Phillipus commented:
I took a look at this. Good job. :-)

My concern is whether GeoReferencedImage and the methods in DiagramUtils belong there or not. Should this class and the methods are quite specific to the HTML Report plugin or whether they have a wider usage?

Phil

Phillipus commented:
Sorry, but after merging the commit I had a closer look at the code and had to undo the commit with another commit to undo the changes.

There are some things that I'm not sure about and need to see how it might impact the rest of the code.

In HTMLReportExporter.java:

```java
private void setOffset(IDiagramModelObject om, int offsetX, int offsetY ) {
    BoundsWithAbsolutePosition b = new BoundsWithAbsolutePosition(om.getBounds());
    b.setOffset(offsetX, offsetY);
    om.setBounds(b);
```

This is setting a new `BoundsWithAbsolutePosition` object at `om.setBounds(b)` and I'm not sure what the impact is of setting this class type which is part of the com.archimatetool.reports.html package. It might have no impact but I need to be sure.

Is there another way to do this so that the object's bounds object is not actually set to a new class?

aschlechter commented:
Hello,

Sorry for the late response.

The GeoreferencedImage and corresponding methods may be useful whereever I
want to know where the model objects are in the image. But we might call
the class ModelReferencedImage...

You are right about possible side effects of the IBounds replacement.

One way to prevent side effects from replacing the original IBounds
implementation with the BoundsWithAbsolutePosition would be to create a
deep copy of the IDiagramModel before replacing the Bounds.

This is what I did in the latest commit.

Greetings,
Tun Schlechter






On Sat, Oct 29, 2016 at 1:21 PM, Phil Beauvoir <notifications@github.com>
wrote:

> Sorry, but after merging the commit I had a closer look at the code and
> had to undo the commit with another commit to undo the changes.
>
> There are some things that I'm not sure about and need to see how it might
> impact the rest of the code.
>
> In HTMLReportExporter.java:
>
> private void setOffset(IDiagramModelObject om, int offsetX, int offsetY ) {
>     BoundsWithAbsolutePosition b = new BoundsWithAbsolutePosition(om.getBounds());
>     b.setOffset(offsetX, offsetY);
>     om.setBounds(b);
>
> This is setting a new BoundsWithAbsolutePosition object at om.setBounds(b)
> and I'm not sure what the impact is of setting this class type which is
> part of the com.archimatetool.reports.html package. It might have no impact
> but I need to be sure.
>
> Is there another way to do this so that the object's bounds object is not
> actually set to a new class?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/archimatetool/archi/pull/154#issuecomment-257086054>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV5LjNOY_Izl13jE39LyrJ2WCa4Kqkt1ks5q4yxNgaJpZM4Kc-C->
> .
>



-- 
*Tun Schlechter*

Abacus Consulting & Solutions S.A.
24, Val Saint André
L-1128 Luxembourg

Tel: +352 621 178 938
Mail: antoine.schlechter@abac.lu

-- 
Note: Privileged/Confidential information may be contained in this message 
and may be subject to legal privilege. Access to this e-mail by anyone 
other than the intended is unauthorised. If you are not the intended 
recipient (or responsible for delivery of the message to such person), you 
may not use, copy, distribute or deliver to anyone this message (or any 
part of its contents ) or take any action in reliance on it. In such case, 
you should destroy this message, and notify us immediately. If you have 
received this email in error, please notify us immediately by e-mail or 
telephone and delete the e-mail from any computer. If you or your employer 
does not consent to internet e-mail messages of this kind, please notify us 
immediately. All reasonable precautions have been taken to ensure no 
viruses are present in this e-mail. As our company cannot accept 
responsibility for any loss or damage arising from the use of this e-mail 
or attachments we recommend that you subject these to your virus checking 
procedures prior to use. The views, opinions, conclusions and other 
informations expressed in this electronic mail are not given or endorsed by 
the company unless otherwise indicated by an authorized representative 
independent of this message.
